### ISSUE TITLE
Need way of resetting state, for subset of mini-batch, within the middle of a batch sequence

### ISSUE THREAD

**Issue by hughperkins** [Element-Research/rnn#101]
Need way of resetting state, for subset of mini-batch, within the middle of a batch sequence.

This means we can have sequences of different lengths in the same mini-batch, and learning will be correct.

Note that backprop would also need to notice that we reset the state, and not backprop earlier than the 'reset' point, for that specific sequence, within the mini-batch.

Alternatively, during backprop, backprop would proceed all the way backwards, and just skip our 'reset' markers.

There seem to be a couple of ways of implementing this, a priorae, meaning, before I've looked at how `rnn` works:
- strip any Sequencer off from our network, train using backwardOnline, and call a method at each step, which takes a vector, whose length matches the batch size, and which contains a `1` for each sequence whose state should be reset at that step, or
- use the Sequencer.  Pass a table of vectors, to the Sequencer, where the meaning of each vector is as for the previous option, ie batch-length vector, where '1' in the vector resets state, for that sequence, at that steps, or
- use the Sequencer.  Have a special input value, that means 'reset state'.  But I think this sounds harder to implement than the other two options?

If we include only the first two options for consideration, the first one is an implementation subset of the second one, ie:
- first we add a method to eg AbstractRecurrent, which wipes the hidden state, for a subset of the sequences, and marks this somehow, so backprop works ok too, then
- second, we can add a way of passing a table of such vectors to Sequencer

**Comment by hughperkins** [comment]
Example for how this might work with an appropriate Sequencer.  Currently, input to Sequencer is a set of sequences:

```
  The quick brown fox
 She sells sea shells
       Black on black
Dark Side of the Moon
```
But the state will be wrong at the start of 'The quick brown fox', since it will have already been updated by 3 time steps.  So, in the revised version, we'd pass in a second set of vectors, where a `1` value means 'reset the hidden state, for this sequence':

```
001000000000000000000
010000000000000000000
000000010000000000000
100000000000000000000
```
(Note that we also have a `1` at the start of 'Dark side of the moon', which means we no longer need to call `:forget()`, and can control where/when we are calling it).

**Comment by nicholas-leonard** [comment]
https://github.com/Element-Research/rnn/pull/46

**Comment by hughperkins** [comment]
Fair point.  Will close this issue for now, since 46 has the advantage of not needing to wipe the state.  I'm still pondering whether it might be useful to be able to wipe the state occasionally, so can send through two partial sequences in the same sub-batch index, but haven't yet come up with a convincing need/use-case for that :-)

**Comment by nicholas-leonard** [comment]
@hughperkins I like the idea of a means to tell the network when to forget at a more granular level. Right now, the [forget()](https://github.com/Element-Research/rnn/blob/master/AbstractRecurrent.lua#L164) method is very simple. It just resets the time-step counter to 1 and prepares the buffers for the next batch. 

In your case, we would need to zero the `gradOutput(t)` and `hidden(t-1)` for the pair (row-id, time-step) (i.e. when a 1 is encountered in the forget matrix). So we could add a`AbstractRecurrent:maskSequence(mask)` function where `mask` is a vector of 0s and 1s, where a 1 indicates that the sequence must be restarted for that example for the current time-step. 

I am reopening this in case we find time to implement this. I think it's worth a try.

**Comment by ZENGXH** [comment]
May I ask currently is there any way to force the LSTM clean up all its states leaving only parameters? Seems that the `:forget()` method from `AbstractRecurrent` does not clean the cells.
Many thanks.

**Comment by nicholas-leonard** [comment]
@ZENGXH If you mean for serialization, you can decorate your model with a Serial module (see dpnn) : 
```
model = nn.Serial(model)
model:mediumSerial()
```
Otherwize, during training, forget does exactly what it is supposed to. It keeps the previous clone so as to not reallocate memory.

**Comment by hughperkins** [comment]
hi.  Cleaning up my https://github.com/issues   If we still want this issue, can you re-open a new issue please, so I'm no longer marked as owner :-)

**Comment by nicholas-leonard** [comment]
@hughperkins Oh forgot about this issue. Actually, this is fixed via MaskZero where zero-maked rows act like a granular forget :)

**Comment by hughperkins** [comment]
cool :-)
